src/realm/query_engine.hpp

```diff
@@ -1684,17 +1686,20 @@ class StringNode : public StringNodeBase {
         TConditionFunction cond;
 
         for (size_t s = start; s < end; ++s) {
-
-            StringData t = get_string(s);
-
             if constexpr (std::is_same_v<TConditionFunction, NotEqual>) {
-                if (m_interned_string) {
-                    const auto id = m_string_interner->lookup(get_string(s));
-                    if (id && m_string_interner->compare(*m_interned_string, *id))
+                if (m_leaf->is_compressed()) {
+                    if (!m_interned_string) {
+                        // The search string has not been interned, so there is no match
+                        return s;
+                    }
+                    // The search string has been interned, so we can compare the string IDs directly
+                    const auto id = m_leaf->get_string_id(s);
+                    if (m_string_interner->compare(*m_interned_string, *id) != 0)
                         return s;
+                    continue;
```

Comment:
This continue is a bit tricky here. Essentially we want to skip getting the string by index `s` at line 1702, since we know that we won't find any match. Maybe a small comment just to be sure we are not going to delete this continue in the future.

Reply:
I have re-arranged the code a bit.